Python/import.c

```diff
@@ -2226,12 +2226,30 @@ PyImport_LoadLazyImport(PyObject *lazy_import)  // was PyImport_ImportDeferred(P
 {
     assert(lazy_import != NULL);
     assert(PyLazyImport_CheckExact(lazy_import));
+    PyThreadState *tstate = _PyThreadState_GET();
+
     PyLazyImport *lz = (PyLazyImport *)lazy_import;
     PyObject *obj = lz->lz_obj;
     if (obj == NULL) {
         obj = _imp_load_lazy_import_impl(lz);
-        if (obj != NULL)
+        if (obj != NULL) {
             lz->lz_obj = obj;
+        }
+        else {
+            PyFrameObject* frame = PyThreadState_GetFrame(tstate);
+            PyCodeObject *code = PyFrame_GetCode(frame);
+            const char *filename = PyUnicode_AsUTF8(code->co_filename);
+            int line = PyFrame_GetLineNumber(frame);
+
+            // only preserve the most recent (innermost) occured LazyImportError
+            if (tstate->curexc_type != PyExc_LazyImportError) {
+                _PyErr_FormatFromCause(PyExc_LazyImportError,
```

Comment:
If you already have `tstate`, you could use `_PyErr_FormatFromCauseTstate` instead and pass it over. Additionally, you could move `_PyThreadState_GET` to the `else` block, so it's used only there.

However, the filename and the line you need to keep should be added to the `PyLazyImport` object, by the time it is created (when executing `IMPORT_NAME` opcode in `ceval.c` or here in `import.c`, when the object is created). You'll need to add a `PyObject` for the filename and an integer for the line number for that (to `PyLazyImport`), and use those here.